lib/ansible/utils/_antsibull_docs_parser/dom.py

```diff
@@ -0,0 +1,330 @@
+    type: "t.Literal[PartType.ENV_VARIABLE]" = PartType.ENV_VARIABLE
+
+
+class ReturnValuePart(NamedTuple):
+    plugin: t.Optional[PluginIdentifier]
+    entrypoint: t.Optional[str]  # present iff plugin.type == 'role'
+    link: t.List[str]
+    name: str
+    value: t.Optional[str]
+    source: t.Optional[str] = None
+    type: "t.Literal[PartType.RETURN_VALUE]" = PartType.RETURN_VALUE
+
+
+class HorizontalLinePart(NamedTuple):
+    source: t.Optional[str] = None
+    type: "t.Literal[PartType.HORIZONTAL_LINE]" = PartType.HORIZONTAL_LINE
+
+
+class ErrorPart(NamedTuple):
+    message: str
+    source: t.Optional[str] = None
+    type: "t.Literal[PartType.ERROR]" = PartType.ERROR
+
+
+AnyPart = t.Union[
+    TextPart,
+    ItalicPart,
+    BoldPart,
+    ModulePart,
+    PluginPart,
+    URLPart,
+    LinkPart,
+    RSTRefPart,
+    CodePart,
+    OptionNamePart,
+    OptionValuePart,
+    EnvVariablePart,
+    ReturnValuePart,
+    HorizontalLinePart,
+    ErrorPart,
+]
+
+
+Paragraph = t.List[AnyPart]
+
+
+class Walker(abc.ABC):
+    """
+    Abstract base class for walker whose methods will be called for parts of a paragraph.
+    """
+
+    @abc.abstractmethod
+    def process_error(self, part: ErrorPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_bold(self, part: BoldPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_code(self, part: CodePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_horizontal_line(self, part: HorizontalLinePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_italic(self, part: ItalicPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_link(self, part: LinkPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_module(self, part: ModulePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_rst_ref(self, part: RSTRefPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_url(self, part: URLPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_text(self, part: TextPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_env_variable(self, part: EnvVariablePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_option_name(self, part: OptionNamePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_option_value(self, part: OptionValuePart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_plugin(self, part: PluginPart) -> None:
+        pass  # pragma: no cover
+
+    @abc.abstractmethod
+    def process_return_value(self, part: ReturnValuePart) -> None:
+        pass  # pragma: no cover
+
+
+class NoopWalker(Walker):
```

Comment:
Is this used anywhere? I don't see any references to it in this PR.

Reply:
If not, could antsibull-docs-parser be refactored so NoopWalker resides in another file (assuming it's even used there at all)? I'd like to avoid unnecessary modifications to antsibull-docs-parser when bundling it in core, but omitting entire files we're not using is reasonable.

Reply:
> Is this used anywhere? I don't see any references to it in this PR.

It is not used here. I vendored the full library, to make updating it easier.

> If not, could antsibull-docs-parser be refactored so NoopWalker resides in another file (assuming it's even used there at all)?

It's not used in the library at all - it is the base class that users of the library should use when implementing their own walker - except if they want to guarantee that they really overwrite all methods. It should be in the same file as `Walker` IMO, thus moving it to another file means that file needs to be included as well.